assignments/02-cli-constructive.md

```diff
@@ -3,77 +3,82 @@
-names like `daily.2021-09-22_0010`. Each of these directories holds a copy of
-the contents of the original directory from a certain point in time.  Take a
-moment to SSH to the server and see for yourself!
-
-The first piece of each snapshot's name---`daily`, in this example---represents
-the frequency at which those snapshots are taken. The second
-part---`2021-09-22_0010`---is the date and time at which it was taken.
-Snapshots that happen more frequently are also deleted more aggressively,
-meaning that the granularity of snapshots goes down the further you go back.
-
-Note that, to prevent overwhelming tools that recursively traverse a directory
-tree, the `.snapshot/` directory only shows up in listings (`ls -a`) in your
-top-level home directory. However, it is also present in every subdirectory if
-you specifically ask for it:
-
+## `build.sh`: making trouble
+
+A lot of systems programming is done in languages such as C, C++, and Rust.
+These languages all have one thing in common: they are predominantly compiled
+before being executed[^interpreters]. This two-step tango means that every time
+you modify your program, you have to compile it before running it.
+
+[^interpreters]: We say *predominantly* because the evaluation strategy is not
+    necessarily part of the language! The language is the abstract concept and
+    the implementation makes it go zoom. There are C and Rust interpreters and
+    there are (for example) Ruby and Python compilers.
+
+That's not fun, especially if it requires some kind of arcane build flags that
+are hard to remember. Or maybe it has a lot of files and your build command is
+getting unwieldy. Either way, you are going to solve your own problems today by
+*writing a build system*.
+
+You need not have written C code ever before starting this part of the
+assignment. We will provide sample code to you so that you can test your build
+system. Later, though, you *will* write some C code and integrate it into your
+build system.
+
+Your job is to write a program in Bash, `build.sh`, that compiles some C
+program. The minimal functional (but not acceptable to submit) solution looks
+something like this:
+
+```sh
+#!/bin/sh
+cc foo.c -o foo
 ```
-$ ls -a Documents/  # No .snapshot/ listed!
-.  ..  do-not-read  top-secret-file
-$ ls Documents/.snapshot/
-daily.2021-09-19_0010             every_four_hours.2021-09-23_1205
-daily.2021-09-20_0010             every_four_hours.2021-09-23_1605
-daily.2021-09-21_0010             every_four_hours.2021-09-23_2005
-daily.2021-09-22_0010             every_four_hours.2021-09-24_0005
-daily.2021-09-23_0010             weekly.2021-09-05_0015
-daily.2021-09-24_0010             weekly.2021-09-12_0015
-every_four_hours.2021-09-23_0405  weekly.2021-09-19_0015
-every_four_hours.2021-09-23_0805
-$ 
+
+If only it were that simple for your assignment. We'll get more into all this
+later in the term, but this program is not ideal: it builds `foo.c` into `foo`
```

Comment:
I might try to reword this, but I'd need to think about it more. Will come back to this comment later.